onnxruntime/core/providers/cpu/nn/shrink.cc

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/cpu/nn/shrink.h"
+
+#include "core/util/math.h"
+#include "core/util/math_cpuonly.h"
+
+namespace onnxruntime {
+ONNX_CPU_OPERATOR_KERNEL(
+    Shrink,
+    9,
+    KernelDefBuilder().TypeConstraint("T", {DataTypeImpl::GetTensorType<float>(),
+                                            DataTypeImpl::GetTensorType<double>(),
+                                            DataTypeImpl::GetTensorType<int64_t>(),
+                                            DataTypeImpl::GetTensorType<uint64_t>(),
+                                            DataTypeImpl::GetTensorType<int32_t>(),
+                                            DataTypeImpl::GetTensorType<uint32_t>(),
+                                            DataTypeImpl::GetTensorType<int16_t>(),
+                                            DataTypeImpl::GetTensorType<uint16_t>(),
+                                            DataTypeImpl::GetTensorType<int8_t>(),
+                                            DataTypeImpl::GetTensorType<uint8_t>(),
+                                            DataTypeImpl::GetTensorType<MLFloat16>(),
+                                            DataTypeImpl::GetTensorType<BFloat16>()}),
+    Shrink);
+
+namespace shrink_internal {
+template <class T>
+inline T ShrinkImpl(const T& val, float bias, float lambd) {
+  // The ONNX spec doesn't take numeric overflow and underflow into account
+  // Implementing the spec as is for now
+  if (val < -lambd) {
+    return T(val + bias);
+  } else if (val > lambd) {
+    return T(val - bias);
+  } else {
+    return T(0);
+  }
+}
+
+void ShrinkMLFloat16(const Tensor* input, Tensor* output, float bias, float lambd) {
+  auto span = gsl::make_span(input->Data<MLFloat16>(), input->Shape().Size());
+  auto output_data = output->template MutableData<MLFloat16>();
+  std::transform(span.cbegin(), span.cend(), output_data, [bias, lambd](const MLFloat16& val) {
+    float fl = math::halfToFloat(val.val);
+    return MLFloat16(math::floatToHalf(ShrinkImpl<float>(fl, bias, lambd)));
+  });
+}
+
+void ShrinkBFloat16(const Tensor* input, Tensor* output, float bias, float lambd) {
+  auto span = gsl::make_span(input->Data<BFloat16>(), input->Shape().Size());
+  auto output_data = output->template MutableData<BFloat16>();
+  std::transform(span.cbegin(), span.cend(), output_data, [bias, lambd](const BFloat16& val) {
+    float fl = val.ToFloat();
+    return BFloat16(ShrinkImpl<float>(fl, bias, lambd));
+  });
+}
+
+}  // namespace shrink_internal
+
+Status Shrink::Compute(OpKernelContext* p_op_kernel_context) const {
+  using namespace shrink_internal;
+
+  auto input = p_op_kernel_context->Input<Tensor>(0);
+  auto output = p_op_kernel_context->Output(0, input->Shape());
+
+  auto dtype = input->DataType();
```

Comment:
Prefer explicitly using 'const' and * or & with 'auto' when applicable so it's clearer what it represents.

Reply:
Thanks. Fixed.